aleo-programs/staking_lite/src/main.leo

```diff
@@ -0,0 +1,376 @@
+    let delegator_balance: u64 = delegator_shares.get(owner);
+    assert(delegator_balance >= withdrawal_shares);
+
+    // Distribute shares for new commission
+    let bonded: u64 = 0u64; // credits.aleo/bonded.get(CORE_PROTOCOL);
+    let current_balance: u64 = total_balance.get(0u8);
+    let current_shares: u64 = total_shares.get(0u8);
+    let rewards: u64 = bonded - current_balance;
+    let commission_rate: u128 = commission_percent.get(0u8);
+    let new_commission: u64 = rewards > 0u64 ? get_commission(rewards as u128, commission_rate) : 0u64;
+    current_balance += rewards - new_commission;
+
+    let new_commission_shares: u64 = calculate_new_shares(current_balance as u128, new_commission as u128, current_shares as u128);
+    let current_commission: u64 = delegator_shares.get_or_use(ADMIN, 0u64);
+    delegator_shares.set(ADMIN, current_commission + new_commission_shares);
+
+    current_shares += new_commission_shares;
+    current_balance += new_commission;
+
+    // Calculate withdrawal amount
+    let withdraw_ratio: u128 = (withdrawal_shares as u128 * PRECISION_UNSIGNED) / current_shares as u128;
+    let withdraw_calculation: u128 = (current_balance as u128 * withdraw_ratio) / PRECISION_UNSIGNED;
+
+    // If the calculated withdrawal amount is greater than total_withdrawal, the excess will stay in the pool
+    assert(withdraw_calculation >= total_withdrawal as u128);
+
+    // Update withdrawals mappings
+    let batch_height: u32 = new_batch ? get_new_batch_height(block.height) : current_batch;
+    current_batch_height.set(0u8, batch_height);
+    let withdrawal: withdrawal_state = withdrawal_state {
+      microcredits: total_withdrawal,
+      claim_block: batch_height
+    };
+    withdrawals.set(owner, withdrawal);
+
+    // Update pending withdrawal
+    let currently_pending: u64 = pending_withdrawal.get(0u8);
+    pending_withdrawal.set(0u8, currently_pending + total_withdrawal);
+
+    // Update total balance
+    total_balance.set(0u8, current_balance - total_withdrawal);
+
+    // Update total shares
+    total_shares.set(0u8, current_shares - withdrawal_shares);
+
+    // Update delegator_shares mapping
+    delegator_shares.set(owner, delegator_balance - withdrawal_shares);
+  }
+
+  inline get_new_batch_height(height: u32) -> u32 {
+    let rounded_down: u32 = (height) / 1_000u32 * 1_000u32;
+    let rounded_up: u32 = rounded_down + 1_000u32;
+    return rounded_up;
+  }
+
+  transition withdraw_claim(amount: u64) {
+    credits.aleo/claim_unbond_public();
+    credits.aleo/transfer_public(self.caller, amount);
+
+    return then finalize(self.caller, amount);
+  }
+
+  finalize withdraw_claim(owner: address, amount: u64) {
+    let withdrawal: withdrawal_state = withdrawals.get(owner);
+    assert(block.height >= withdrawal.claim_block);
+    assert_eq(withdrawal.microcredits, amount);
+
+    // Allow unbonding now that credits have been claimed
+    current_batch_height.remove(0u8);
```

Comment:
Something to check -- is it cheaper to set mappings to 0 rather than removing them?